File: apps/server-admin/src/lib/ddns/providers/edgeone-cname.ts
import type {
  DDNSProviderContext,
  DDNSProviderDefinition,
  DDNSUpdateResult,
} from "../types";
import { ddnsProviderT, normalizeDomain } from "./helpers";
import {
  EDGEONE_OVERSEAS_ACCESS_MODE_FIELD,
  requestEdgeOneJson,
} from "./edgeone-shared";

const edgeoneCnameT = (
  key: string,
  params?: Record<string, string | number | boolean | null | undefined>,
) => ddnsProviderT("edgeone_cname", key, params);

type EdgeOneOriginDetail = {
  HostHeader?: string | null;
  Origin?: string;
  OriginType?: string;
};

type EdgeOneAccelerationDomain = {
  DomainName?: string;
  OriginDetail?: EdgeOneOriginDetail | null;
};

type EdgeOneDescribeAccelerationDomainsResponse = {
  AccelerationDomains?: EdgeOneAccelerationDomain[];
  TotalCount?: number;
};

export const edgeoneCnameProvider: DDNSProviderDefinition = {
  name: "edgeone_cname",
  label: edgeoneCnameT("label"),
  fields: [
    {
      key: "secret_id",
      label: "SecretId",
      type: "text",
      placeholder: "AKID...",
      required: true,
    },
    {
      key: "secret_key",
      label: "SecretKey",
      type: "password",
      placeholder: edgeoneCnameT("fields.secret_key.placeholder"),
      required: true,
    },
    {
      key: "zone_id",
      label: "Zone ID",
      type: "text",
      placeholder: "zone-xxxxxxxx",
      required: true,
      description: edgeoneCnameT("fields.zone_id.description"),
    },
    {
      key: "domain",
      label: edgeoneCnameT("fields.domain.label"),
      type: "text",
      placeholder: "home.example.com",
      required: true,
      description: edgeoneCnameT("fields.domain.description"),
    },
    {
      key: EDGEONE_OVERSEAS_ACCESS_MODE_FIELD,
      label: edgeoneCnameT("fields.overseas_access.label"),
      type: "select",
      required: false,
      options: [
        {
          label: edgeoneCnameT("fields.overseas_access.options.off"),
          value: "off",
        },
        {
          label: edgeoneCnameT("fields.overseas_access.options.blockOverseas"),
          value: "block_overseas",
        },
      ],
      description: edgeoneCnameT("fields.overseas_access.description"),
    },
    {
      key: "endpoint",
      label: "API Endpoint",
      type: "text",
      placeholder: "https://teo.tencentcloudapi.com",
      required: false,
      description: edgeoneCnameT("fields.endpoint.description"),
    },
    {
      key: "region",
      label: "Region",
      type: "text",
      placeholder: edgeoneCnameT("fields.region.placeholder"),
      required: false,
      description: edgeoneCnameT("fields.region.description"),
    },
  ],
};

async function edgeOneCnameRequest<T>(
  context: DDNSProviderContext,
  action: string,
  payload: Record<string, unknown>,
): Promise<T> {
  const { config } = context;
  const secretId = config.secret_id?.trim();
  const secretKey = config.secret_key?.trim();
  if (!secretId || !secretKey) {
    throw new Error(edgeoneCnameT("configIncomplete"));
  }

  return requestEdgeOneJson<T>(context, action, payload);
}

function resolveDesiredOrigin(
  ipv4: string | null,
  ipv6: string | null,
): { family: "ipv4" | "ipv6"; value: string } {
  if (ipv4 && ipv6) {
    throw new Error(edgeoneCnameT("singleAddressOnly"));
  }

  if (ipv4) {
    return { family: "ipv4", value: ipv4 };
  }

  if (ipv6) {
    return { family: "ipv6", value: ipv6 };
  }

  throw new Error(edgeoneCnameT("noIpAvailable"));
}

function isValidCustomHostHeader(value: string | undefined): boolean {
  const host = normalizeDomain(value || "");
  if (!host) {
    return false;
  }

  if (
    host.includes("/") ||
    host.includes(":") ||
    host.includes("[") ||
    host.includes("]") ||
    host.includes("*") ||
    /\s/.test(host) ||
    /^https?:\/\//i.test(host)
  ) {
    return false;
  }

  if (host.length > 253) {
    return false;
  }

  return host.split(".").every((label) => {
    return (
      label.length > 0 &&
      label.length <= 63 &&
      !label.startsWith("-") &&
      !label.endsWith("-") &&
      /^[a-z0-9-]+$/i.test(label)
    );
  });
}

function isHostHeaderFormatError(error: unknown): boolean {
  const message = error instanceof Error ? error.message : String(error);
  return (
    message.includes("InvalidHostHeaderFormat") ||
    message.includes("HostHeaderInvalid")
  );
}

function buildOriginInfoPayload(
  desiredOrigin: { family: "ipv4" | "ipv6"; value: string },
  hostHeader?: string,
): Record<string, string> {
  return {
    OriginType: "IP_DOMAIN",
    Origin: desiredOrigin.value,
    ...(hostHeader ? { HostHeader: hostHeader } : {}),
  };
}

export async function edgeoneCnameUpdate(
  context: DDNSProviderContext,
  ipv4: string | null,
  ipv6: string | null,
): Promise<DDNSUpdateResult> {
  const { config } = context;
  const secretId = config.secret_id?.trim();
  const secretKey = config.secret_key?.trim();
  const zoneId = config.zone_id?.trim();
  const domain = normalizeDomain(config.domain || "");
  if (!secretId || !secretKey || !zoneId || !domain) {
    return {
      success: false,
      message: edgeoneCnameT("configIncomplete"),
    };
  }

  let desiredOrigin: ReturnType<typeof resolveDesiredOrigin>;
  try {
    desiredOrigin = resolveDesiredOrigin(ipv4, ipv6);
  } catch (error) {
    return {
      success: false,
      message: error instanceof Error ? error.message : String(error),
    };
  }

  const list =
    await edgeOneCnameRequest<EdgeOneDescribeAccelerationDomainsResponse>(
      context,
      "DescribeAccelerationDomains",
      {
        ZoneId: zoneId,
        Offset: 0,
        Limit: 20,
        Match: "all",
        Filters: [
          {
            Name: "domain-name",
            Values: [domain],
            Fuzzy: false,
          },
        ],
      },
    );

  const existing = (list.AccelerationDomains || []).find((item) => {
    return normalizeDomain(item.DomainName || "") === domain;
  });

  if (!existing) {
    return {
      success: false,
      message: edgeoneCnameT("domainNotFound", { domain }),
    };
  }

  const originType = (existing.OriginDetail?.OriginType || "")
    .trim()
    .toUpperCase();
  if (originType && originType !== "IP_DOMAIN") {
    return {
      success: false,
      message: edgeoneCnameT("unsupportedOriginType", { originType }),
    };
  }

  const currentOrigin = existing.OriginDetail?.Origin?.trim() || "";
  const rawHostHeader = existing.OriginDetail?.HostHeader?.trim();
  const hostHeader = isValidCustomHostHeader(rawHostHeader)
    ? normalizeDomain(rawHostHeader!)
    : undefined;
  const ignoredInvalidHostHeader = Boolean(rawHostHeader) && !hostHeader;

  if (currentOrigin === desiredOrigin.value) {
    return {
      success: true,
      message: edgeoneCnameT("originUnchanged"),
      ipv4Updated: desiredOrigin.family === "ipv4",
      ipv6Updated: desiredOrigin.family === "ipv6",
    };
  }

  try {
    await edgeOneCnameRequest(context, "ModifyAccelerationDomain", {
      ZoneId: zoneId,
      DomainName: domain,
      OriginInfo: buildOriginInfoPayload(desiredOrigin, hostHeader),
    });
  } catch (error) {
    if (!hostHeader || !isHostHeaderFormatError(error)) {
      throw error;
    }

    await edgeOneCnameRequest(context, "ModifyAccelerationDomain", {
      ZoneId: zoneId,
      DomainName: domain,
      OriginInfo: buildOriginInfoPayload(desiredOrigin),
    });
  }

  return {
    success: true,
    message: ignoredInvalidHostHeader
      ? edgeoneCnameT("successWithInvalidHostHeaderIgnored")
      : edgeoneCnameT("success"),
    ipv4Updated: desiredOrigin.family === "ipv4",
    ipv6Updated: desiredOrigin.family === "ipv6",
  };
}

File: apps/server-admin-rs/src/ddns/routes/providers/edgeone_cname.rs
use super::*;

pub(in crate::ddns::routes) fn edgeone_cname_catalog_entry() -> Value {
    let mut value = provider(
        "edgeone_cname",
        "EdgeOne CNAME",
        vec![
            field("secret_id", "SecretId", "text", "AKID...", true),
            field("secret_key", "SecretKey", "password", "SecretKey", true),
            field("zone_id", "Zone ID", "text", "zone-xxxxxxxx", true),
            field("domain", "Domain", "text", "home.example.com", true),
            select_field(
                DDNS_EDGEONE_OVERSEAS_ACCESS_FIELD,
                "Overseas access control",
                false,
                vec![("Off", "off"), ("Block overseas IPs", "block_overseas")],
            ),
            field(
                "endpoint",
                "API Endpoint",
                "text",
                "https://teo.tencentcloudapi.com",
                false,
            ),
            field("region", "Region", "text", "", false),
        ],
    );
    if let Some(object) = value.as_object_mut() {
        object.insert(
            "capabilities".to_string(),
            json!({ "addressMode": "single_address" }),
        );
    }
    value
}

pub(in crate::ddns::routes) async fn update_edgeone_cname(
    translator: &Translator,
    config: &HashMap<String, String>,
    http_options: &DDNSHttpClientOptions,
    ipv4: Option<&str>,
    ipv6: Option<&str>,
) -> anyhow::Result<DDNSProviderUpdateResult> {
    let secret_id = config_value(config, "secret_id");
    let secret_key = config_value(config, "secret_key");
    let zone_id = config_value(config, "zone_id");
    let domain = normalize_domain(&config_value(config, "domain"));
    if secret_id.is_empty() || secret_key.is_empty() || zone_id.is_empty() || domain.is_empty() {
        return Ok(provider_failure(ddns_text(
            translator,
            "providers.edgeone_cname.configIncomplete",
            &[],
        )));
    }
    let desired = match (ipv4, ipv6) {
        (Some(_), Some(_)) => {
            return Ok(provider_failure(ddns_text(
                translator,
                "providers.edgeone_cname.singleAddressOnly",
                &[],
            )));
        }
        (Some(value), None) => ("ipv4", value),
        (None, Some(value)) => ("ipv6", value),
        (None, None) => {
            return Ok(provider_failure(ddns_text(
                translator,
                "providers.edgeone_cname.noIpAvailable",
                &[],
            )));
        }
    };
    let client = ddns_http_client(translator, http_options)?;
    let list = edgeone_request(
        translator,
        &client,
        config,
        &secret_id,
        &secret_key,
        "DescribeAccelerationDomains",
        json!({
            "ZoneId": zone_id,
            "Offset": 0,
            "Limit": 20,
            "Match": "all",
            "Filters": [{
                "Name": "domain-name",
                "Values": [domain],
                "Fuzzy": false
            }]
        }),
    )
    .await?;
    let existing = list
        .get("AccelerationDomains")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .find(|item| {
            normalize_domain(
                item.get("DomainName")
                    .and_then(Value::as_str)
                    .unwrap_or_default(),
            ) == domain
        })
        .cloned();
    let Some(existing) = existing else {
        return Ok(provider_failure(ddns_text(
            translator,
            "providers.edgeone_cname.domainNotFound",
            &[("domain", domain.clone())],
        )));
    };
    let origin_detail = existing.get("OriginDetail").unwrap_or(&Value::Null);
    let origin_type = origin_detail
        .get("OriginType")
        .and_then(Value::as_str)
        .unwrap_or_default()
        .trim()
        .to_ascii_uppercase();
    if !origin_type.is_empty() && origin_type != "IP_DOMAIN" {
        return Ok(provider_failure(ddns_text(
            translator,
            "providers.edgeone_cname.unsupportedOriginType",
            &[("originType", origin_type)],
        )));
    }
    let current_origin = origin_detail
        .get("Origin")
        .and_then(Value::as_str)
        .unwrap_or_default()
        .trim();
    if current_origin == desired.1 {
        return Ok(DDNSProviderUpdateResult {
            success: true,
            message: ddns_text(translator, "providers.edgeone_cname.originUnchanged", &[]),
        });
    }
    let raw_host_header = origin_detail
        .get("HostHeader")
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .map(str::to_string);
    let host_header = raw_host_header
        .as_deref()
        .filter(|value| is_valid_edgeone_host_header(value))
        .map(normalize_domain);
    let ignored_invalid_host_header = raw_host_header.is_some() && host_header.is_none();
    let modify_result = edgeone_request(
        translator,
        &client,
        config,
        &secret_id,
        &secret_key,
        "ModifyAccelerationDomain",
        json!({
            "ZoneId": zone_id,
            "DomainName": domain,
            "OriginInfo": edgeone_cname_origin_info(desired.1, host_header.as_deref())
        }),
    )
    .await;
    if let Err(error) = modify_result {
        if host_header.is_none() || !is_edgeone_host_header_format_error(&error) {
            return Err(error);
        }
        edgeone_request(
            translator,
            &client,
            config,
            &secret_id,
            &secret_key,
            "ModifyAccelerationDomain",
            json!({
                "ZoneId": zone_id,
                "DomainName": domain,
                "OriginInfo": edgeone_cname_origin_info(desired.1, None)
            }),
        )
        .await?;
    }
    Ok(DDNSProviderUpdateResult {
        success: true,
        message: ddns_text(
            translator,
            if ignored_invalid_host_header {
                "providers.edgeone_cname.successWithInvalidHostHeaderIgnored"
            } else {
                "providers.edgeone_cname.success"
            },
            &[],
        ),
    })
}

pub(in crate::ddns::routes) fn edgeone_cname_origin_info(
    origin: &str,
    host_header: Option<&str>,
) -> Value {
    let mut value = json!({
        "OriginType": "IP_DOMAIN",
        "Origin": origin
    });
    if let Some(host_header) = host_header.filter(|value| !value.is_empty()) {
        insert_json_field(&mut value, "HostHeader", json!(host_header));
    }
    value
}

pub(in crate::ddns::routes) fn is_edgeone_host_header_format_error(error: &anyhow::Error) -> bool {
    let message = error.to_string();
    message.contains("InvalidHostHeaderFormat") || message.contains("HostHeaderInvalid")
}
